source/src/localisation/cs_TrackableParser.cpp

```diff
@@ -124,136 +223,371 @@ void TrackableParser::handleScannedDevice(scanned_device_t* device) {
+		// command totalSize only includes the size of the cuckoo filter and its metadata, not the runtime data yet.
 		parsingFilter = allocateParsingFilter(cmd_data->filterId, cmd_data->totalSize);
 
 		if (parsingFilter == nullptr) {
 			// failed to handle command, no space.
+			LOGTrackableParserWarn("parsing filter allocation failed");
 			return ERR_NO_SPACE;
 		}
-
-		// initialize runtime data.
-		parsingFilter->runtimedata.filterId = cmd_data->filterId;
-		parsingFilter->runtimedata.crc      = 0;
-
-		// meta data and filter data will be memcpy'd from chunks,
-		// no need to copy those.
 	}
 
-	// WARNING(Arend): we're not doing corruption checks here yet.
-	// E.g.: when totalSize changes for a specific filterId before
-	//       a commit is reached, we have a parsingFilter allocated
-	//       but it's of the wrong size.
-	// Edit(Arend):
-	// 		As we keep a list of pointers _parsingFilters, as well
-	// 		as the end pointer to the byte array, this information
-	// 		is implicitly available even if multiple filters are
-	// 		uploaded in chunks concurrently. Just ptrdiff them.
+	// by now, the filter exists, is clean, and the incoming chunk is verified for totalSize consistency.
 
 	// chunk index starts counting from metadata onwards (ignoring runtimedata)
 	uint8_t* parsingFilterBase_ptr  = reinterpret_cast<uint8_t*>(&(parsingFilter->metadata));
 	uint8_t* parsingFilterChunk_ptr = parsingFilterBase_ptr + cmd_data->chunkStartIndex;
 
-	if (parsingFilterChunk_ptr + cmd_data->chunkSize > _filterBuffer + FILTER_BUFFER_SIZE) {
-		// chunk would overwrite end of filterbuffer.
-		// (Unlikely to happen, previous guards should have caught this.)
-		return ERR_NO_SPACE;
-	}
-
 	// apply filter chunk, counting chunk index from metadata onwards:
 	std::memcpy(parsingFilterChunk_ptr, cmd_data->chunk, cmd_data->chunkSize);
 
 	return ERR_SUCCESS;
 }
 
-cs_ret_code_t TrackableParser::handleRemoveFilterCommand(
-		trackable_parser_cmd_remove_filter_t* cmd_data) {
-	// TODO(Arend): implement later.
-	return ERR_NOT_IMPLEMENTED;
+cs_ret_code_t TrackableParser::handleRemoveFilterCommand(trackable_parser_cmd_remove_filter_t* cmd_data) {
+	startProgress();
+
+	return deallocateParsingFilter(cmd_data->filterId) ? ERR_SUCCESS : ERR_NOT_FOUND;
 }
 
 cs_ret_code_t TrackableParser::handleCommitFilterChangesCommand(
 		trackable_parser_cmd_commit_filter_changes_t* cmd_data) {
-	// TODO(Arend): implement later.
-	// - compute and check all filter sizes
-	// - compute and check all filter crcs
-	// - compute and check(?) master crc
-	// - persist all filters
-	// - unset in progress flag (async?)
-	// - broadcast update to the mesh
-	return ERR_NOT_IMPLEMENTED;
+
+	if (checkFilterSizeConsistency()) {
+		return ERR_WRONG_STATE;
+	}
+
+	computeCrcs();
+
+	uint16_t masterHash = masterCrc();
+	if (cmd_data->masterCrc != masterHash) {
+		LOGTrackableParserWarn("Master Crc did not match (%x != %x, filterModificationInProgress stays true.",
```

Comment:
It's going to bite you at some point when the hex value has decimal numbers only, and you compare it to a decimal value.
```suggestion
		LOGTrackableParserWarn("Master Crc did not match (0x%x != 0x%x, filterModificationInProgress stays true.",
```

Reply:
Painful truth, wouldn't be the first time...